README
# Copyright (C) 2007-2013 Codership Oy <info@codership.com>

Mailing list:
https://groups.google.com/forum/?fromgroups=#!forum/codership-team

DISCLAIMER:
===========
This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY, to the extent permitted by law; without even the
implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

See COPYING for license details.


ABOUT:
======
glbd is a simple TCP connection balancer made with scalability and
performance in mind. It was inspired by pen, but unlike pen its
functionality is limited only to balancing generic TCP connections.

Features:
 - list of backend servers is configurable in runtime.
 - supports server "draining", i.e. does not allocate new connections to
   server, but does not kill existing ones, waiting for them to end
   gracefully.
 - can use epoll API provided by Linux version 2.6 and higher for ultimate
   routing performance.
 - glb is multithreaded, so it can utilize multiple CPU cores. Also, if your
   OS does not support epoll API, consider using several threads even on a
   single core machine as it will lessen poll() overhead proportionally and
   can improve overall performance by a factor of 2 or more.
 - optional watchdog module can monitor destinations and adjust routing table
   automatically.

libglb is a shared library that provides 0-effort connection balancing to
any Linux application that uses standard libc connect() call by overloading
that function in runtime. No other program functionality is affected and no
program modification or recompilation is needed. See below for details.


BALANCING POLICIES:
===================
GLB supports five balancing "policies":

a) "least connected" - new connection will be directed to the server with
   least connections (corrected for server "weight"). This policy is default.

b) "round-robin" - each new connection is routed to the next destination
   in the list in circular order.

c) "single" - all connections are routed to a single server with the highest
   weight available. All routing will stick to that server until it fails or
   a server with a strictly higher weight is introduced.

d) "random" - connections are distributed randomly between the servers.

e) "source tracking" - connections originating from the same address are
   directed to the same server. For details about this policy see below.

-T|--top option was introduced in GLB 0.9.2. It restricts all balancing
policies to a subset of destinations with top weight. E.g. if there are servers
configured with weight 1 and 2, all balancing will happen only between servers
with weight 2 as long as at least one of them is available.


MAXIMUM CONCURRENT CONNECTIONS:
===============================
Maximum connections that can be opened via glbd simultaneously depends on the
system open files limit and is 493 for a standard limit of 1024. If needed glbd
will attempt to increase open files limit as much as allowed by the effective
user privileges. For unprivileged user it is normally 4096 which results
in max 2029 connections.
On Linux open files limit may be checked with 'ulimit -n' and if necessary
increased in /etc/security/limits.conf.


COMMAND LINE OPTIONS:
=====================
See output of the --help option.


RUNTIME MANAGEMENT:
===================
Runtime management can be done either through FIFO file, or network socket.
By default network socket is not opened; address and port to listen at must
be explicitly specified with -c option.

To add/modify/delete backend server:
send server specification in the form <IP address>:<port>[:weight] where
weight is an integer to the daemon. Connections are distributed proportionally
to the weight. Default weight is 1. Weight of 0 means drain the server.
Negative weight means delete the server completely (all connections to that
server are closed immediately). This works both on socket connection and on
FIFO file.

To see the stats:
send "getstat" command to the daemon. This works only on socket connection
since it implies response.

Example (here glbd is listening at 127.0.0.1:4444):
$ echo "192.168.0.1:3307:5" | nc -q 1 127.0.0.1 4444
OK
$ echo "192.168.0.2:3307:5" | nc -q 1 127.0.0.1 4444
OK
$ echo "getinfo" | nc -q 1 127.0.0.1 4444
Router:
----------------------------------------------------
   Address       : weight   usage  conns
192.168.0.1:3307 :  5.000   0.000    0
192.168.0.2:3307 :  5.000   0.000    0
----------------------------------------------------
Destinations: 2, total connections: 0 

'usage' here is some dimensionless metric of how much destination is staffed
with connections (relative to weight). Ranges from 0 (totally unused) to 1.0
(very busy). Router tries to keep 'usage' equal on all destinations.


ADDRESS CONVENTIONS:
====================
All network addresses are specified in the form IP|HOSTNAME:PORT:WEIGHT.
Depending on the context some parts can be optional, in that case they can be
omitted. For example address to listen for client connections can be specified
either as HOSTNAME:PORT or just PORT. In the latter case glbd will listen for
client connections on all interfaces. Backend servers can be specified as
HOSTNAME1,HOSTNAME2,HOSTNAME3. In that case incoming port number will be used
for PORT value and 1 will be used for WEIGHT value.


PERFORMANCE STATISTICS:
=======================
GLB allows to query raw performance statistics through control socket using
'getstat' command. The client can use these data to obtain useful information,
e.g. average number of reads per poll() call.

$ echo "getstat" | nc -q 1 127.0.0.1 4444
in: 6930 out: 102728 recv: 109658 / 45 send: 109658 / 45 conns: 0 / 4 poll: 45 / 0 / 45 elapsed: 1.03428

Statistics line consists of fields separated by spaces for ease of parsing in
scripts. A few description fields are added to assist in human reading. Value
fields are all even and go as follows:

  2  - number of bytes received on incoming interface (client requests)
  4  - number of bytes sent from incoming interface (server responses)
  6  - number of bytes passed through recv() call
  8  - number of recv() calls
  10 - number of bytes passed through send() call (should be equal to p.6)
  12 - number of send() calls
  14 - number of created connections
  16 - number of concurrent connections
  18 - number of read-ready file descriptors returned by poll()/epoll_wait()
  20 - number of write-ready file descriptors returned by poll()/epoll_wait()
  22 - number of times poll()/epoll_wait() triggered
  24 - time elapsed since last statistics report (seconds)

All values except for 16 and 24 are totals accumulated since the last report.
In order to obtain some variable rate it must be divided by the elapsed time.
On 32-bit architectures the values are stored in 4-byte integers and can
overflow if too much time elapsed, so the first statistics report in the series
may need to be discarded.


SOURCE TRACKING CAPABILITY:
===========================
GLB features simple source tracking capability where connections originating
from one address can be routed to the same destination, chosen randomly
among available destinations according to their weights. One limitation of such
scheme is that when the destination list changes, destination choice for new
connections changes as well while established connections naturally stay
unchanged. Also when a destination is marked unavailable, all connections that
would be routed to it will fail over to another, randomly chosen destination.
When the original target becomes available, all new connections will be routed
back to it.

In other words source tracking should be considered a best effort feature and
will work best for short-lived connections.


WATCHDOG:
=========
(NOTE: this is a work in progress and neither functionality nor terminology is
       final.)

Without the watchdog GLB can check destination availability only via the
ability to establish TCP connection to destination. For most use cases TCP
connectivity is an insufficient check. E.g. for HTTP server it is important to
be able to serve web pages and for DBMS server - to be able to execute queries.
Watchdog module implements asynchronous monitoring of destinations via backends
made to check particular service availability.

Watchdog is enabled with a -w|--watchdog option. Its parameter is a string
containing the backend ID string, optionally followed by a colon and backend
configuration options. Example:

$ glbd -w exec:"mysql.sh -utest -ptestpass" -t 2 3306 192.168.0.1 192.168.0.2

This initializes "exec" backend that executes external programs. In this case
it will execute mysql.sh script (can be found in the files/ directory,
must be placed in PATH) with provided parameters to check MySQL servers
at192.168.0.1 and 192.168.0.2 (in parallel threads). To pass server
address to a script, 'exec' backend will insert host:port string as the first
argument after a script/command name, so the actual command that would be
executed in this case looks like:

"mysql.sh 192.168.0.1:3306 -utest -ptestpass"

Check interval is set with -i|--interval parameter (fractional seconds, 1.0).


DESTINATION DISCOVERY:
======================
If destinations can supply information about other members of the cluster
it can be used to automatically populate watchdog destination list if
-D|--discover option is supplied. Currently only MySQL/Galera nodes are known
to do this, so it is a Galera-only option.


RUNNING GLBD AS A "SERVICE":
============================
See README in the files/ directory.


USING LIBGLB:
=============
Using libglb requires 2 environment variables to be set:

LD_PRELOAD=<path-to-libglb>

GLB_OPTIONS='options string'
    Allows to specify GLB options to libglb the same way as command line
    parameters for glbd. It is limited however in that options and their values
    cannot contain whitespaces and commas and short options cannot be
    concatenated (i.e. '-qri10' should go as '-q -r -i 10'). Parsing errors and
    options unsupported by libglb (like -d) will be silently ignored.
    In particular, watchdog option cannot be specified this way.
    (See GLB_WATCHDOG environment variable below)

Example:

$ LD_PRELOAD=src/.libs/libglb.so \
GLB_OPTIONS="--random 3306 192.168.0.1 192.168.0.2 192.168.0.3" \
mysql -uroot -prootpass -h127.0.0.1 -P3306

Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 76
Server version: 5.5.28 Source distribution, wsrep_24dev.7.r3830
...


Additional libglb parameters:

In case GLB_OPTIONS is not sufficient (e.g. watchdog option needs to be
specified) below there is a list of additional environment variables (which
take precedence over GLB_OPTIONS). Note however, that all these options except
GLB_WATCHDOG are considered deprecated at the moment.

GLB_WATCHDOG=<watchdog specification>
    Interpreted the same way as --watchdog parameter.

GLB_TARGETS=H1[:P1[:W1]],[H2[:P2[:W2]],...]
    Is a comma-separated list of target servers among which the client
    connections must be distributed.

GLB_BIND=<addr>
    Where <addr> is interpreted the same way as LISTEN_ADDR parameter for glbd.
    Whenever application attempts to initiate connection to this address, the
    request will be intercepted and connection established to one of the
    servers from GLB_TARGETS list according to balancing rules.

GLB_POLICY=single|random|source
    Default libglb balancing policy is "round-robin", "single", "random" and
    "source tracking" policies can be specified with GLB_POLICY variable.

(The meaning of GLB_POLICY=source in this case is that all connections from
this client will be routed to the same random destination, and fail over to
another if the primary destination fails. Thus client-server affinity is
achieved, however load from many clients will be spread over all available
destinations.)

GLB_CONTROL=[IP:]PORT
    Interpreted the same way as --control parameter of glbd. Application will
    open a socket at a specified address to listen to control commands. Due
    to library functionality limited only to connect() call, no traffic
    statistics or connection count is maintained, so "getstat" command is a
    noop and "getinfo" only prints out a routing table.



ChangeLog
2013-11-25 alex 
- Fixed broken failover from failed destination when using asynchronous
  connects (default).
- added -l|--linger option to disable socket lingering after connection close.
- must be builadble on FreeBSD now.
- various small compilation fixes and such.
  Version 1.0.1

2013-03-15 alex 
- Fixed issue with LD_PRELOAD and GLB environment variables being inherited
  by the processes spawned by watchdog (and, consequently, spawning more).
  Version 1.0.0

2013-03-04 alex 
- Fixed libglb initialization which was thread-unsafe and unoptimal.
- Fixed GLB_OPTIONS parsing in libglb, which didn't actually work.
- Added logging capability to libglb and --verbose option support.
- Cleaned up logging.
- Fixed -D|--discovery option, which didn't have effect.
  Version 1.0.0rc1

2013-01-21 alex 
- Added automatic Galera node discovery (-D|--discover option).
- Added GLB_OPTIONS env. variable to set misc. options for libglb.
- Added a sample curl script for checking HTTP servers.
- Fixed a bunch of leaks and uninitialized values shown by valgrind.
  Version 1.0.0beta

2013-01-18 alex 
  Merged with version 0.9.2:
- New option -T|--top: all policies use only top-weighted destinations for
  routing. E.g. in the presence of destinations with weights 1 and 2 only
  the latter will be used to route connections to until all of them fail.
- New policy 'single': all connections are directed to only one destination
  chosen by highest weight.
- Added SO_KEEPALIVE option to server side sockets to avoid stalled connections
  Version 1.0.0alpha4

2013-01-10 alex 
- Merged in asynchronous connection processing from 0.9.1 and other missing
  fixes from the trunk.
  Version 1.0.0alpha3

2012-12-23 alex 
- Implemented initial watchdog framework that asynchronously monitors
  destinations using server-specific probes and updates routing table
  accordingly. (Destination discovery is not supported yet)
- Implemented basic watchdog module that spawns destination probes via shell,
  allowing various probing scripts to be written.
- Added ability to use watchdog probes to probe destinations immediately before
  connect.
  Version 1.0.0alpha2

2012-12-19 alex 
- Fixed division by 0 when destination list is empty introduced by
  "round-robin" policy support.
  Version 0.9.0

2012-12-09 alex 
- added "round-robin" balancing policy.
- made "random" and "source tracking" policies to obey destination weight.
- libglb: a shared library that adds balancing capabilities to any application
  without recompilation.
  Version 0.9.0beta

2012-12-05 alex 
- *HOTFIX*: fixed slow MySQL connection initiation due to TCP_DEFER_ACCEPT
  enabled by default. Now it is an option.
- added "random" load balancing policy.
  Version 0.8.2

2012-12-01 alex 
- Fixed privileges issue when using service script and default FIFO location.
- Implemented maximum connections option which both limits maximum client
  connections and also tries to increase OS open files limit if necessary.
- Added MAX_CONN parameter to start/stop script.
  Version 0.8.1

2012-11-30 alex 
- Fixed obscure connection loss when too many connections are initiated at once.
- Fixed segfault when -s option was specified.
- Added TCP_NODELAY option to sockets (on by default)
- Added start/stop script and config template to run glbd as a "service".
- Implemented basic IP-based source tracking mechanism
  (same IP clients directed to the same server).
  Version 0.8.0

2012-05-17 alex 
- Fixed glb_sockaddr_t initialization + compilation cleanups.
  Version 0.7.5

2010-02-15 teemu 
- Fixed usage counting bug when usage could become negative when all
  destinations fail simultaneously.
  Version 0.7.4

2009-11-04 alex 
- Compilation cleanups for gcc-4.4.1

2009-09-11 alex 
- Router lock is no longer held while connecting. However destination table
  modification is still blocked.

2009-07-25 alex 
- Logging cleanups: converted the last of fprintfs and perrors to use logger.
  Version 0.7.3

2009-06-07 alex 
- Fixed:
  * Off-by-one bug which threw control thread into a tight loop when using FIFO
    control method.
  * listener thread was still using select().
  Version 0.7.2

2009-04-18 alex 
- Fixed:
  * control thread was still using select() to listen to requests (FD_SETSIZE
    descriptor limit).
  * reverted "fix" which opened control socket by default for security and
    usability considerations.
  Version 0.7.1.

2009-04-16 alex 
- Added machine parsable stats output on 'getstat' command
- Fixed:
  * control socket was not opened by default
  * usage display if weight was not 0 or 1
  * real numbers were not accepted for destination weight
  Version 0.7.0.

2009-04-15 alex 
- Refactored to use poll/epoll API instead of select(). This allows for more
  than FD_SETSIZE/2 concurrent connections and is considerably faster in case
  of epoll.

2009-03-12 alex 
- Cosmetic changes to the output: usage now ranges strictly from 0.0
  (unused) to 1.0 (overloaded). Weight 0.0 automatically means usage 1.0,
  before that it was 'nan'. Version 0.6.3.

2009-02-10 alex 
- Fixed non-working destination failover, failed destinations are retried
  every 2 seconds. Version 0.6.2.

2009-01-07 alex 
- Fixed silly destination list parsing bug. Version 0.6.1.

2008-09-14 alex 
- Fixed a problem with default port when adding destinations during runtime.
- Changed from comma-separated destination list to space-separated
  destination list, hence bumped minor version.

2008-07-15 alex 
- Initial release.


src/glb_cmd.c
/*
 * Copyright (C) 2008-2013 Codership Oy <info@codership.com>
 *
 * $Id: glb_cmd.c 156 2013-08-23 08:24:56Z vlad $
 */

#include "glb_cmd.h"
#include "glb_opt.h"
#include "glb_limits.h"
#include "glb_socket.h"

#include <errno.h>
#include <assert.h>
#include <ctype.h> // isspace()
#include <stdlib.h>

extern char* optarg;

// Defaults relevant to CLI
static const char cmd_inc_addr_default[]  = "0.0.0.0";
static const char cmd_ctrl_addr_default[] = "127.0.0.1";


static void
cmd_parse_options (int argc, char* argv[], glb_cnf_t* cnf)
{
    int   opt = 0;
    int   opt_idx = 0;
    char* endptr;

    // parse options
    while ((opt = getopt_long (argc, argv, "DKL:STVYabc:dfhi:lm:nt:rsvw:x:",
                               glb_options, &opt_idx)) != -1) {
        switch (opt) {
        case GLB_OPT_DISCOVER:
            cnf->discover = true;
            break;
        case GLB_OPT_KEEPALIVE:
            cnf->keepalive = false;
            break;
        case GLB_OPT_LATENCY_COUNT:
            cnf->lat_factor = strtol (optarg, &endptr, 10);
            if ((*endptr != '\0' && !isspace(*endptr)) || errno ||
                cnf->lat_factor < 0) {
                fprintf (stderr, "Bad latency count value: %s. "
                         "Non-negative integer expected.\n", optarg);
                exit (EXIT_FAILURE);
            }
            break;
        case GLB_OPT_SINGLE:
            cnf->policy = GLB_POLICY_SINGLE; // implies GLB_OPT_TOP
        case GLB_OPT_TOP:
            cnf->top = true;
            break;
        case GLB_OPT_VERSION:
            glb_print_version (stdout);
            if (argc == 2) exit(0);
            break;
        case GLB_OPT_SYNCHRONOUS:
            cnf->synchronous = true;
            break;
        case GLB_OPT_DEFER_ACCEPT:
            cnf->defer_accept = true;
            break;
        case GLB_OPT_ROUND_ROBIN:
            cnf->policy = GLB_POLICY_ROUND;
            break;
        case GLB_OPT_CONTROL:
            if (glb_parse_addr (&cnf->ctrl_addr, optarg, cmd_ctrl_addr_default))
                exit (EXIT_FAILURE);
            cnf->ctrl_set = true;
            break;
        case GLB_OPT_DAEMON:
            cnf->daemonize = true;
            break;
        case GLB_OPT_FIFO:
            cnf->fifo_name = optarg;
            break;
        case '?':
        case GLB_OPT_HELP:
            glb_cmd_help(stdout, argv[0]);
            exit (EXIT_FAILURE);
            break;
        case GLB_OPT_INTERVAL:
            cnf->interval = glb_time_from_double(strtod (optarg, &endptr));
            if ((*endptr != '\0' && !isspace(*endptr)) || errno ||
                cnf->interval <= 0) {
                fprintf (stderr, "Bad check interval value: %s. "
                         "Positive real number expected.\n", optarg);
                exit (EXIT_FAILURE);
            }
            break;
        case GLB_OPT_LINGER:
            cnf->linger = true;
            break;
        case GLB_OPT_MAX_CONN:
            cnf->max_conn = strtol (optarg, &endptr, 10);
            if ((*endptr != '\0' && !isspace(*endptr)) || errno) {
                fprintf (stderr, "Bad max_conn value: %s. Integer expected.\n",
                         optarg);
                exit (EXIT_FAILURE);
            }
            break;
        case GLB_OPT_NODELAY:
            cnf->nodelay = false;
            break;
        case GLB_OPT_N_THREADS:
            cnf->n_threads = strtol (optarg, &endptr, 10);
            if ((*endptr != '\0' && !isspace(*endptr)) || errno) {
                fprintf (stderr, "Bad n_threads value: %s. Integer expected.\n",
                         optarg);
                exit (EXIT_FAILURE);
            }
            break;
        case GLB_OPT_RANDOM:
            cnf->policy = GLB_POLICY_RANDOM;
            break;
        case GLB_OPT_SRC_TRACKING:
            cnf->policy = GLB_POLICY_SOURCE;
            break;
        case GLB_OPT_VERBOSE:
            cnf->verbose = true;
            break;
        case GLB_OPT_WATCHDOG:
            cnf->watchdog = optarg;
            break;
        case GLB_OPT_EXTRA_POLLS:
            cnf->extra = glb_time_from_double(strtod (optarg, &endptr));
            if ((*endptr != '\0' && !isspace(*endptr)) || errno ||
                cnf->extra < 0) {
                fprintf (stderr, "Bad extra value: %s. "
                         "Non-negative real number expected.\n", optarg);
                exit (EXIT_FAILURE);
            }
            break;
        default:
            fprintf (stderr, "Option '%s' (%d) not supported yet. Ignoring.\n",
                     glb_options[opt_idx].name, opt);
        }
    }
}

void
glb_cmd_help (FILE* out, const char* progname)
{
    fprintf (out,
             "Usage:\n  %s [OPTIONS] LISTEN_ADDRESS "
             "[DESTINATION_LIST]\nOPTIONS:\n", progname);
    fprintf (out,
             "  -h|--help                 this help message.\n");
    fprintf (out,
             "  -a|--defer-accept         "
             "enable TCP_DEFER_ACCEPT on the listening socket\n"
             "                            (default: disabled).\n");
    fprintf (out,
             "  -b|--round                "
             "round-robin destination selection policy.\n");
    fprintf (out,
             "  -c|--control [HOST:]PORT  "
             "listen for control requests on this address.\n");
    fprintf (out,
             "  -d|--daemon               run as a daemon.\n");
    fprintf (out,
             "  -f|--fifo <fifo name>     "
             "name of the FIFO file for control.\n");
    fprintf (out,
             "  -i|--interval D.DDD       "
             "how often to probe destinations for liveness\n"
             "                            (fractional seconds, default 1.0).\n"
        );
    fprintf (out,
             "  -l|--linger               "
             "*DISABLE* sockets lingering in TIME_WAIT state after\n"
             "                            close().\n");
    fprintf (out,
             "  -m|--max_conn N           "
             "maximum allowed number of client connections\n"
             "                            (OS dependent).\n");
    fprintf (out,
             "  -n|--nodelay              "
             "*DISABLE* TCP_NODELAY socket option\n"
             "                            (default: enabled).\n");
    fprintf (out,
             "  -r|--random               "
             "route connections to randomly selected destination.\n");
    fprintf (out,
             "  -s|--source               "
             "turn on source tracking: route connections from one\n"
             "                            source to the same destination.\n");
    fprintf (out,
             "  -t|--threads N            "
             "number of working threads (connection pools).\n");
    fprintf (out,
             "  -v|--verbose              turn on verbose reporting.\n");
    fprintf (out,
             "  -w|--watchdog SPEC_STR    watchdog specification.\n");
    fprintf (out,
             "  -x|--extra D.DDD          "
             "perform extra destination poll on connection attempt\n"
             "                            "
             "if the previous poll happened more than D.DD seconds\n"
             "                            ago.\n"
             "                            "
             "(default: 0.0 - extra polling disabled)\n");
    fprintf (out,
             "  -D|--discover             "
             "use watchdog results to discover and set new\n"
             "                            destinations.\n"
             "                            "
             "(Currently only Galera nodes supply such info.)\n");
    fprintf (out,
             "  -K|--keepalive            "
             "*DISABLE* SO_KEEPALIVE socket option on server-side\n"
             "                            sockets (default: enabled).\n");
    fprintf (out,
             "  -L|--latency <samples>    "
             "when using latency reported by watchdog probes for\n"
             "                            "
             "destination weight adjustment, how many samples to\n"
             "                            average latency over.\n"
             "                            "
             "(default: 0 - not using reported latency for weight\n"
             "                            adjustment)\n");
    fprintf (out,
             "  -S|--single               "
             "direct all connections to a single destination\n"
             "                            "
             "with top weight.\n");
    fprintf (out,
             "  -T|--top                  "
             "balance only between destinations with top weight.\n");
    fprintf (out,
             "  -V|--version              print program version.\n");
    fprintf (out,
             "  -Y                        "
             "connect synchronously (one-at-a-time).\n");
    fprintf (out, "LISTEN_ADDRESS:\n"
             "  [IP:]PORT                 "
             "where to listen for incoming TCP connections at.\n"
             "                            "
             "(without IP part - bind to all interfaces)\n"
             );
    fprintf (out, "DESTINATION_LIST:\n"
             "  [H1[:P1[:W1]]] [H2[:P2[:W2]]]... "
             " - a space-separated list of destinations\n"
             "                            in the form address:port:weight.\n");
    fprintf (out, "SPEC_STR:\n"
             "  BACKEND_ID[:BACKEND_SPECIFIC_STRING], "
             "e.g. exec:'<command line>'\n");
    exit (EXIT_FAILURE);
}


glb_cnf_t*
glb_cmd_parse (int argc, char* argv[])
{
    glb_cnf_t*   cnf = glb_cnf_init(); // initialize to defaults
    const char** dst_list = NULL;
    uint16_t     inc_port;

    if (!cnf) exit (EXIT_FAILURE);

    // parse options
    cmd_parse_options (argc, argv, cnf);

    // first non-option argument
    if (optind >= argc) {
        fprintf (stderr, "Missing required argument: LISTEN_ADDR.\n");
        glb_cmd_help(stderr, argv[0]);
        exit (EXIT_FAILURE);
    }

    // parse obligatory incoming address
    if (glb_parse_addr (&cnf->inc_addr, argv[optind], cmd_inc_addr_default)) {
        exit (EXIT_FAILURE);
    }
    inc_port = glb_sockaddr_get_port (&cnf->inc_addr);

#if 0 // don't open socket by default for security considerations.
    // if control address was not specified
    if (!cnf->ctrl_set) {
        char port[6] = { 0, };
        snprintf (port, 5, "%hu", inc_port + 1);
        if (glb_parse_addr (&cnf->ctrl_addr, port, cmd_ctrl_addr_default))
            exit (EXIT_FAILURE);
        cnf->ctrl_set = true;
    }
#endif

    // if number of threads was not specified
    if (!cnf->n_threads) cnf->n_threads = 1;

    if (cnf->max_conn > glb_get_conn_limit()) {
        int res = glb_set_conn_limit (cnf->max_conn);
        if (res > 0) cnf->max_conn = res;
    }

    if (cnf->daemonize) cnf->verbose = false;

    // parse destination list
    if (++optind < argc) dst_list = (const char**) &(argv[optind]);
    assert (argc >= optind);

    return glb_parse_dst_list (dst_list, argc - optind, inc_port, cnf);
}


